File: 4. Binary Tree/7_tree_path_finder.java
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;

class Node<T> {
    T val;
    Node<T> left;
    Node<T> right;

    public Node(T val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

class Source {

    public static <T> List<T> returnPath(Node<T> root, T target) {

        if (root == null) {
            return null;
        }

        if (root.val == target) {
            List<T> list = new ArrayList<>();
            list.add(target);
            return list; // array list is mutable (List is not)
        }

        List<T> leftPath = returnPath(root.left, target);

        if (leftPath != null) {
            leftPath.add(root.val); // bcs it's an arraylist I can add to it
            return leftPath;
        }

        List<T> rightPath = returnPath(root.right, target);

        if (rightPath != null) {
            rightPath.add(root.val); // bcs it's an arraylist I can add to it
            return rightPath;
        }

        return null;

    }

    public static <T> List<T> pathFinder(Node<T> root, T target) {
        List<T> path = returnPath(root, target); // O(n)
        if (path == null) {
            return null;
        } else {
            Collections.reverse(path);
            return path;
        }

    }
}

File: 4. Binary Tree/7_tree_path_finder.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def return_path(root, target):
    if root is None:
        return None

    if root.val == target:
        return [target]

    left_path = return_path(root.left, target)
    if left_path is not None:
        left_path.append(root.val)
        return left_path

    right_path = return_path(root.right, target)
    if right_path is not None:
        right_path.append(root.val)
        return right_path

    return None

def path_finder(root, target):
    path = return_path(root, target)
    if path is None:
        return None
    path.reverse()
    return path
